src/TuxState.js

```diff
@@ -0,0 +1,650 @@
+    for (var key in propsToExtend) {
+      if (currentState[key]) {
+        //if we find any key that maches a key in the current state, set matchedKey to true
+        matchedKey = true;
+      }
+    }
+    //throw error if no keys match any keys in the current state
+    invariant(matchedKey, "At least one outer key must match an outer key in the current state. Use setState if you only wish to add new keys and not change existing keys.");
+    //build new object with all keys in the current state plus any additional keys in propsToExtend, overwriting any matching keys with those in propsToExtend
+    var newState = assign({}, currentState, propsToExtend);
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //pushState FUNCTION: push the value of the deepest keys in the passed in object to the corresponding deepest keys in the current state. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToPush OBJECT: required object argument with deepest keys being a single value to push to the array in the corresponding key in the current state
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  pushState: function (propsToPush, callback) {
+    //throw error if propsToPush is undefined or not an object
+    invariantArgCheck(propsToPush);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToPush, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that current state at this key is an array
+      invariantArrayCheck(currentStateAtKey);
+      //check that there is only one value being pushed in
+      invariantValueCheck(newPropsAtKey);
+      //push new value into the corresponding array in the current state
+      currentStateAtKey.push(newPropsAtKey);
+      newState[key] = currentStateAtKey;
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //popState FUNCTION: match the deepest keys in the passed in object with the corresponding deepest keys in the current state and pop an item off the array in the corresponding key. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToPush OBJECT: required object argument with deepest keys being a boolean
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  popState: function (propsToPop, callback) {
+    //throw error if propsToPop is undefined or not an object
+    invariantArgCheck(propsToPop);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
```

Comment:
I don't think we need to invoke `assign` here because we are already making a shallow copy of state at the first line in `buildNewState`.